### PR TITLE
ASQ native integration sample

### PR DESCRIPTION
**Pull request by SeanFeldman**

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@mauroservienti, I've tried to put together a new sample for ASQ transport using Core v8 but it seems that whatever is on MyGet is not ready yes. Specifically, the transport aspect is failing (see commented-out lines in the code for ASQN_10). Also, the transport is failing with the following exception: 

<details>
<summary>Exception info</summary>

System.Exception
  HResult=0x80131500
  Message=The selected persistence doesn't have support for subscription storage. Select another persistence or disable the publish functionality using transportConfiguration.DisablePublishing()
  Source=NServiceBus.Core
  StackTrace:
   at NServiceBus.Features.MessageDrivenSubscriptions.Setup(FeatureConfigurationContext context) in /_/src/NServiceBus.Core/Routing/MessageDrivenSubscriptions/MessageDrivenSubscriptions.cs:line 67
   at NServiceBus.Features.FeatureActivator.FeatureInfo.InitializeFrom(FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureActivator.cs:line 233
   at NServiceBus.Features.FeatureActivator.ActivateFeature(FeatureInfo featureInfo, List`1 featuresToActivate, FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureActivator.cs:line 190
   at NServiceBus.Features.FeatureActivator.SetupFeatures(FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureActivator.cs:line 53
   at NServiceBus.FeatureComponent.Initalize(FeatureConfigurationContext featureConfigurationContext) in /_/src/NServiceBus.Core/Features/FeatureComponent.cs:line 28
   at NServiceBus.EndpointCreator.Configure() in /_/src/NServiceBus.Core/EndpointCreator.cs:line 63
   at NServiceBus.EndpointCreator.Create(SettingsHolder settings, Configuration hostingConfiguration) in /_/src/NServiceBus.Core/EndpointCreator.cs:line 28
   at NServiceBus.HostCreator.<CreateWithInternallyManagedContainer>d__1.MoveNext() in /_/src/NServiceBus.Core/Hosting/HostCreator.cs:line 57
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at NServiceBus.Endpoint.<Start>d__1.MoveNext() in /_/src/NServiceBus.Core/Endpoint.cs:line 29
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Program.<Main>d__0.MoveNext() in C:\github\nsb\docs.particular.net\samples\azure\native-integration-asq\ASQN_10\Receiver\Program.cs:line 41
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Program.<Main>()

  This exception was originally thrown at this call stack:
    NServiceBus.Features.MessageDrivenSubscriptions.Setup(NServiceBus.Features.FeatureConfigurationContext) in MessageDrivenSubscriptions.cs
    NServiceBus.Features.FeatureActivator.FeatureInfo.InitializeFrom(NServiceBus.Features.FeatureConfigurationContext) in FeatureActivator.cs
    NServiceBus.Features.FeatureActivator.ActivateFeature(NServiceBus.Features.FeatureActivator.FeatureInfo, System.Collections.Generic.List<NServiceBus.Features.FeatureActivator.FeatureInfo>, NServiceBus.Features.FeatureConfigurationContext) in FeatureActivator.cs
    NServiceBus.Features.FeatureActivator.SetupFeatures(NServiceBus.Features.FeatureConfigurationContext) in FeatureActivator.cs
    NServiceBus.FeatureComponent.Initalize(NServiceBus.Features.FeatureConfigurationContext) in FeatureComponent.cs
    NServiceBus.EndpointCreator.Configure() in EndpointCreator.cs
    NServiceBus.EndpointCreator.Create(NServiceBus.Settings.SettingsHolder, NServiceBus.HostingComponent.Configuration) in EndpointCreator.cs
    NServiceBus.HostCreator.CreateWithInternallyManagedContainer(NServiceBus.EndpointConfiguration) in HostCreator.cs
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
    ...
    [Call Stack Truncated]

</details>

The suggestion seems to be outdated
> The selected persistence doesn't have support for subscription storage. Select another persistence or disable the publish functionality using transportConfiguration.DisablePublishing()

Is this expected at this point and I can merge the sample as-is or there's a newer version of the transport that could be used? Thanks.

File: samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        Console.Title = endpointName;
+
+        var endpointConfiguration = new EndpointConfiguration(endpointName);
+        endpointConfiguration.UseSerialization<NewtonsoftSerializer>();
+        endpointConfiguration.EnableInstallers();
+
+        var transport = new AzureStorageQueueTransport("UseDevelopmentStorage=true");
+        // TODO: ASQ transport is not yet ready, Core v8 work is still in progress
+        // transport.DisablePublishing();
+        // transport.DelayedDelivery().DisableDelayedDelivery();
```
Comment (mauroservienti) [inline samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs:21]: To disable native delayed deliveries use the following constructor overload:

```
var transport = new AzureStorageQueueTransport("UseDevelopmentStorage=true", useNativeDelayedDeliveries: false );
```

**Comment by mauroservienti** [comment]
@timbussmann is this something that needs to be addressed in Core? Neither v9 nor the current version of the transport has anything to do with the `DisablePublishing()` thing that seems to be entirely managed by Core. As far as I can understand to me it work we need a new intention method in Core. As of now, there is the following:

```csharp
public static void DisablePublishing<T>(this TransportExtensions<T> transportExtensions) where T : TransportDefinition, IMessageDrivenSubscriptionTransport
{
   transportExtensions.Settings.Set(MessageDrivenSubscriptions.EnablePublishingSettingsKey, false);
}
```

We probably need something like:

```csharp
public static void DisablePublishing(this IMessageDrivenSubscriptionTransport transport)
{
   //...
}
```

ASQ implements `IMessageDrivenSubscriptionTransport`

File: samples/azure/native-integration-asq/ASQN_10/NativeSender/Program.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+            var nativeMessage = new NativeMessage
+            {
+                Content = $"Hello from native sender @ {DateTimeOffset.Now}"
+            };
+
+            var serializedMessage = JsonConvert.SerializeObject(nativeMessage);
+            var bytes = Encoding.UTF8.GetBytes(serializedMessage);
+
+            await queueClient.SendMessageAsync(new BinaryData(bytes)).ConfigureAwait(false);
```
Comment (danielmarbach) [inline samples/azure/native-integration-asq/ASQN_10/NativeSender/Program.cs:37]: Aren't there overloads that don't require to specify binary data?

File: samples/azure/native-integration-asq/ASQN_10/Receiver/Receiver.csproj [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  <ItemGroup>
+    <PackageReference Include="Newtonsoft.Json" Version="12.*" />
+    <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="3.0.0-*" />
+    <PackageReference Include="NServiceBus.Transport.AzureStorageQueues" Version="10.0.0-PullRequest0510.393" />
+    <ProjectReference Include="..\Shared\Shared.csproj" />
+  </ItemGroup>
+
+  <ItemGroup Label="Force a later version of a transitive dependency">
+    <PackageReference Include="System.Net.NameResolution" Version="4.3.0" />
```
Comment (danielmarbach) [inline samples/azure/native-integration-asq/ASQN_10/Receiver/Receiver.csproj:17]: Is this still required?

File: samples/azure/native-integration-asq/ASQN_9/NativeSender/Program.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+            var nativeMessage = new NativeMessage
+            {
+                Content = $"Hello from native sender @ {DateTimeOffset.Now}"
+            };
+
+            var serializedMessage = JsonConvert.SerializeObject(nativeMessage);
+            var bytes = Encoding.UTF8.GetBytes(serializedMessage);
+
+            await queueClient.SendMessageAsync(new BinaryData(bytes)).ConfigureAwait(false);
```
Comment (danielmarbach) [inline samples/azure/native-integration-asq/ASQN_9/NativeSender/Program.cs:37]: Same comment as above

File: samples/azure/native-integration-asq/ASQN_9/Receiver/Program.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+        #region Native-message-mapping
+
+        transport.UnwrapMessagesWith(message => new MessageWrapper
+        {
+            Id = message.MessageId,
+            Body = message.Body.ToArray(),
+            Headers = new Dictionary<string, string>
+            {
+                { "NServiceBus.EnclosedMessageTypes", typeof(NativeMessage).FullName }
```
Comment (danielmarbach) [inline samples/azure/native-integration-asq/ASQN_9/Receiver/Program.cs:32]: Should we directly use the headers constant given that we reference NSB already?

File: samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        #region Native-message-mapping
+
+        transport.MessageUnwrapper = message => new MessageWrapper
+        {
+            Id = message.MessageId,
+            Body = message.Body.ToArray(),
+            Headers = new Dictionary<string, string>
+            {
+                {"NServiceBus.EnclosedMessageTypes", typeof(NativeMessage).FullName}
```
Comment (danielmarbach) [inline samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs:31]: Should we directly use the headers constant given that we reference NSB already?

**Comment by timbussmann** [comment]
> @timbussmann is this something that needs to be addressed in Core? Neither v9 nor the current version of the transport has anything to do with the DisablePublishing() thing that seems to be entirely managed by Core.

Not sure what you mean. DisablePublishing is only an option when using MessageDriven pub/sub and it tells core to not use it. The transports are not involved in pub/sub at all anyway in these scenarios. The option still exists, it's just on the RoutingSettings now, so:

```
var settings = endpointConfig.UseTransport(new Blabla());
settings.DisablePublishing();
```

(this doesn't work with the "shim" API though, that might be something that would have to be added?)

File: samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs (hunk 1 above)
Comment (SeanFeldman) [reply samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs:21]: Thank you. Too bad `transport.DelayedDelivery().DisableDelayedDelivery();` does not point in the right direction. Someone might overlook the constructor overload for the transport just like I did.

File: samples/azure/native-integration-asq/ASQN_10/NativeSender/Program.cs (hunk 2 above)
Comment (SeanFeldman) [reply samples/azure/native-integration-asq/ASQN_10/NativeSender/Program.cs:37]: Yes. Good point, I've overlooked it.

File: samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs (hunk 1 above)
Comment (SeanFeldman) [reply samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs]: @mauroservienti, this doesn't seem to work as expected. The `RecoverabilityComponent` still throws.

<details>
<summary>Exception</summary>

System.Exception
  HResult=0x80131500
  Message=Delayed retries are not supported when the transport does not support delayed delivery.
  Source=NServiceBus.Core
  StackTrace:
   at NServiceBus.RecoverabilityComponent.GetDelayedRetryConfig() in /_/src/NServiceBus.Core/Recoverability/RecoverabilityComponent.cs:line 140
   at NServiceBus.RecoverabilityComponent.Initialize(Configuration receiveConfiguration, Configuration hostingConfiguration, TransportSeam transportSeam) in /_/src/NServiceBus.Core/Recoverability/RecoverabilityComponent.cs:line 56
   at NServiceBus.EndpointCreator.Configure() in /_/src/NServiceBus.Core/EndpointCreator.cs:line 65
   at NServiceBus.EndpointCreator.Create(SettingsHolder settings, Configuration hostingConfiguration) in /_/src/NServiceBus.Core/EndpointCreator.cs:line 28
   at NServiceBus.HostCreator.<CreateWithInternallyManagedContainer>d__1.MoveNext() in /_/src/NServiceBus.Core/Hosting/HostCreator.cs:line 57
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at NServiceBus.Endpoint.<Start>d__1.MoveNext() in /_/src/NServiceBus.Core/Endpoint.cs:line 29
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Program.<Main>d__0.MoveNext() in C:\github\nsb\docs.particular.net\samples\azure\native-integration-asq\ASQN_10\Receiver\Program.cs:line 39
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at Program.<Main>()

  This exception was originally thrown at this call stack:
    NServiceBus.RecoverabilityComponent.GetDelayedRetryConfig() in RecoverabilityComponent.cs
    NServiceBus.RecoverabilityComponent.Initialize(NServiceBus.ReceiveComponent.Configuration, NServiceBus.HostingComponent.Configuration, NServiceBus.TransportSeam) in RecoverabilityComponent.cs
    NServiceBus.EndpointCreator.Configure() in EndpointCreator.cs
    NServiceBus.EndpointCreator.Create(NServiceBus.Settings.SettingsHolder, NServiceBus.HostingComponent.Configuration) in EndpointCreator.cs
    NServiceBus.HostCreator.CreateWithInternallyManagedContainer(NServiceBus.EndpointConfiguration) in HostCreator.cs
    System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
    System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(System.Threading.Tasks.Task)
    System.Runtime.CompilerServices.ConfiguredTaskAwaitable<TResult>.ConfiguredTaskAwaiter.GetResult()
    NServiceBus.Endpoint.Start(NServiceBus.EndpointConfiguration) in Endpoint.cs
    ...
    [Call Stack Truncated]

</details>

Comment (SeanFeldman) [reply samples/azure/native-integration-asq/ASQN_10/Receiver/Program.cs]: nm, I had to disable the delayed recoverability.

File: samples/azure/native-integration-asq/ASQN_10/Receiver/Receiver.csproj (hunk 3 above)
Comment (SeanFeldman) [reply samples/azure/native-integration-asq/ASQN_10/Receiver/Receiver.csproj:17]: Yes.

**Review by danielmarbach — APPROVED** [review]

**Comment by mauroservienti** [comment]
> (this doesn't work with the "shim" API though, that might be something that would have to be added?)

Agreed @timbussmann

**Review by mauroservienti — APPROVED** [review]

**Pull request merged**